lighthouse-core/audits/seo/hreflang.js

```diff
@@ -76,33 +88,55 @@ class Hreflang extends Audit {
    * @param {LH.Artifacts} artifacts
    * @return {LH.Audit.Product}
    */
-  static audit(artifacts) {
-    /** @type {Array<{source: string|{type: 'node', snippet: string}}>} */
+  static audit({LinkElements}) {
+    /** @type {Array<{source: string|{type: 'node', snippet: string}, reason: string}>} */
     const invalidHreflangs = [];
 
-    for (const link of artifacts.LinkElements) {
-      if (link.rel !== 'alternate') continue;
-      if (!link.hreflang || isValidHreflang(link.hreflang)) continue;
-      if (link.source === 'body') continue;
+    const auditableLinkElements = LinkElements.filter(linkElement => {
+      const isAlternate = linkElement.rel === 'alternate';
+      const hasHreflang = linkElement.hreflang;
+      const isInBody = linkElement.source === 'body';
+
+      return isAlternate && hasHreflang && !isInBody;
+    });
+
+    for (const link of auditableLinkElements) {
+      const reasons = [];
+      let source;
+
+      if (!isExpectedLanguageCode(link.hreflang)) {
+        reasons.push(UIStrings.unexpectedLanguage);
+      }
+
+      if (!isFullyQualified(link.hrefRaw.toLowerCase())) {
+        reasons.push(UIStrings.notFullyQualified);
+      }
 
       if (link.source === 'head') {
-        invalidHreflangs.push({
-          source: {
-            type: 'node',
-            snippet: `<link rel="alternate" hreflang="${link.hreflang}" href="${link.href}" />`,
-          },
-        });
+        source = {
+          type: 'node',
+          snippet: `<link rel="alternate" hreflang="${link.hreflang}" href="${link.hrefRaw}" />`,
+        };
       } else if (link.source === 'headers') {
+        source = `Link: <${link.hrefRaw}>; rel="alternate"; hreflang="${link.hreflang}"`;
+      }
+
+      if (!source || !reasons.length) continue;
+
+      for (const reason of reasons) {
         invalidHreflangs.push({
-          source: `Link: <${link.href}>; rel="alternate"; hreflang="${link.hreflang}"`,
+          source,
```

Comment:
Seem to be getting a type error of:

`Type '{ type: string; snippet: string; }' is not assignable to type 'string | { type: "node"; snippet: string; }` 🤔️